### PR TITLE
New authorization context login method revocation system

### LINKED ISSUE #5736: Add token revocation to auth_context logged users when mapping rules change
Hello team,

We should improve our current token blacklist system to revoke those tokens from users logged through an authorization_context whose permissions may have changed due to changes in the security mapping rules.

Currently, when a policy, role, or user is modified, a list of affected users is calculated and all of their tokens are revoked. However, when an user is logged through an authorization_context, its permissions may also change when a new mapping rule is added or deleted from a role.

We should ensure the tokens are revoked correctly in this situation.

### Tasks:

- [x] Update [check_relationships()](https://github.com/wazuh/wazuh/blob/develop/framework/wazuh/core/security.py#L50) function to correctly interpret affected users logged through `auth_context` when a rule changes
- [x] Add `auth_context` logging test cases to [jwt_invalidation](https://github.com/wazuh/wazuh-qa/tree/master/tests/system/test_jwt_invalidation) ansible tests (https://github.com/wazuh/wazuh-qa/issues/840)
- [x] Add `auth_context` integration tests

### PR DESCRIPTION
**Pull request by AdriiiPRodri**
|Related issue|
|---|
|#5736|

Hi team,

This PR closes #5736. In this PR we have added a new system to revoke tokens granted through the authorization contexts based authentication method. In addtion, we have added missing pretty parameters.

### Unittests results

```bash
adriiiprodri@wazuh python3 -m pytest -xs api/api/test/test_authentication.py
========================================================= test session starts =========================================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.7.0, pluggy-0.13.1
rootdir: /home/adriiiprodri/Desktop/git/wazuh/api
plugins: html-2.0.1, mock-1.10.0, asyncio-0.12.0, aiohttp-0.3.0, cov-2.8.1, testinfra-5.0.0, tavern-1.0.0, metadata-1.8.0
collected 10 items

api/api/test/test_authentication.py ..........

=================================================== 10 passed, 2 warnings in 0.55s ====================================================


adriiiprodri@wazuh python3 -m pytest -xs framework/wazuh/rbac/tests/
========================================================= test session starts =========================================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.7.0, pluggy-0.13.1
rootdir: /home/adriiiprodri/Desktop/git/wazuh/framework
plugins: html-2.0.1, mock-1.10.0, asyncio-0.12.0, aiohttp-0.3.0, cov-2.8.1, testinfra-5.0.0, tavern-1.0.0, metadata-1.8.0
collected 218 items

framework/wazuh/rbac/tests/test_auth_context.py ..
framework/wazuh/rbac/tests/test_decorators.py .........................................................................................................
framework/wazuh/rbac/tests/test_default_configuration.py ...............................................
framework/wazuh/rbac/tests/test_orm.py .....................................................
framework/wazuh/rbac/tests/test_preprocessor.py ...........

============================================= 218 passed, 3 warnings in 189.12s (0:03:09) =============================================
```

### Integrated results

```bash
adriiiprodri@wazuh python3 run_tests.py -rbac both -k security
Collected tests [6]:
test_rbac_black_security_endpoints.tavern.yaml, test_rbac_white_security_endpoints.tavern.yaml, test_security_DELETE_endpoints.tavern.yaml, test_security_GET_endpoints.tavern.yaml, test_security_POST_endpoints.tavern.yaml, test_security_PUT_endpoints.tavern.yaml


test_rbac_black_security_endpoints.tavern.yaml 
	 23 passed, 122 warnings

test_rbac_white_security_endpoints.tavern.yaml 
	 24 passed, 122 warnings

test_security_DELETE_endpoints.tavern.yaml 
	 14 passed, 87 warnings

test_security_GET_endpoints.tavern.yaml 
	 17 passed, 128 warnings

test_security_POST_endpoints.tavern.yaml 
	 6 passed, 62 warnings

test_security_PUT_endpoints.tavern.yaml 
	 8 passed, 51 warnings
```

Regards

### CODE REVIEW

File: api/test/integration/test_authorization_contexts_login_method.tavern.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
+---
```
Comment (vicferpoy) [inline api/test/integration/test_authorization_contexts_login_method.tavern.yaml:1]: Change test name. Suggestions:
- test_authorization_context.tavern.yaml
- test_authorization_context_login.tavern.yaml

File: api/test/integration/test_authorization_contexts_login_method.tavern.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        extra_kwargs:
+          expected_roles: []
+
+
+  - name: Login with an authorization context an a valid user
+    request:
+      verify: False
+      url: "{protocol:s}://{host:s}:{port:d}/security/user/authenticate/run_as"
+      headers:
+        Authorization: "Basic {basic_auth:s}"
```
Comment (vicferpoy) [inline api/test/integration/test_authorization_contexts_login_method.tavern.yaml:31]: We are using the same `user:pass` as above. Is this correct? Take a look at the var `basic_auth` in `common.yaml`

File: framework/wazuh/rbac/orm.py [hunk 3]
```diff
@@ -171,6 +171,37 @@ def to_dict(self):
                 'is_valid_until': self.is_valid_until}
 
 
+class RolesTokenBlacklist(_Base):
+    """
+    This table contains the roles with an invalid token and for how long
+    The information stored is:
+        role_id: Affected role id
+        nbf_invalid_until: The tokens that has an nbf prior to this timestamp will be invalidated
```
Comment (vicferpoy) [inline framework/wazuh/rbac/orm.py:179]:
```suggestion
        nbf_invalid_until: The tokens that have an nbf prior to this timestamp will be invalidated
```

File: framework/wazuh/rbac/orm.py [hunk 4]
```diff
@@ -369,13 +400,15 @@ class TokenManager:
     all the methods needed for the token blacklist administration.
     """
 
-    def is_token_valid(self, user_id: int, token_nbf_time: int):
+    def is_token_valid(self,token_nbf_time: int, user_id: int = None, role_id: int = None):
```
Comment (vicferpoy) [inline framework/wazuh/rbac/orm.py:403]:
```suggestion
    def is_token_valid(self, token_nbf_time: int, user_id: int = None, role_id: int = None):
```

File: framework/wazuh/rbac/orm.py [hunk 5]
```diff
@@ -453,19 +508,27 @@ def delete_all_expired_rules(self):
 
         Returns
         -------
-        List of removed user's rules
+        List of removed users and roles rules
```
Comment (vicferpoy) [inline framework/wazuh/rbac/orm.py:511]:
```suggestion
        List of removed user and role rules
```

**Review by vicferpoy — CHANGES_REQUESTED** [review]

**Review by vicferpoy — APPROVED** [review]

**Review by davidjiglesias — APPROVED** [review]

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**